apps/browser-extension-wallet/src/lib/scripts/background/services/cardanoTokenPrices.ts

```diff
@@ -0,0 +1,116 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+import { Subject } from 'rxjs';
+import { storage } from 'webextension-polyfill';
+import { TokenPrices, Status } from '../../types';
+import { Wallet } from '@lace/cardano';
+import { Cardano } from '@cardano-sdk/core';
+import { config } from '@src/config';
+
+/** Shortcut for the token prices subject type. */
+type TokenPriceSubject = Subject<{ tokens: TokenPrices } & Status>;
+
+/** The subset of token data from CoinGecko relevant for Lace to show token prices. */
+type PriceData = [priceInAda: number, priceVariationPercentage24h: number];
+
+/** The structure to store token data and when given data for a token was fetched optimized to have the thinnest size once stringified. */
+type FetchedPriceData = [lastFetchTime: number] | [lastFetchTime: number, priceData: PriceData];
+
+const CACHE_KEY = 'cardano-token-prices';
+const { TOKEN_PRICE_CHECK_INTERVAL } = config();
+
+/**
+ * Given the total amount of token prices data is quite small, rather than fetching it from cache every time it is required,
+ * the entire data is cached in RAM by `priceData` to make read accesses as cheap as possible.
+ *
+ * Write accesses are centralized through `updatePriceData` to ensure data consistency.
+ */
+let priceData: Record<string, FetchedPriceData>;
+let priceSubject$: TokenPriceSubject;
+
+const emitPrices = () =>
+  priceSubject$.next({
+    status: 'fetched',
+    tokens: new Map(
+      Object.entries(priceData)
+        .filter(([, data]) => data)
+        .map(([assetId, [lastFetchTime, data]]) => {
+          const asset = assetId as Wallet.Cardano.AssetId;
+
+          if (!data) return [asset, { lastFetchTime }];
+
+          const [priceInAda, priceVariationPercentage24h] = data;
+
+          return [asset, { lastFetchTime, price: { priceInAda, priceVariationPercentage24h } }];
+        })
+    )
+  });
+
+const updatePriceData = (assetId: string, data?: PriceData) => {
+  const now = Date.now();
+
+  priceData[assetId] = data ? [now, data] : [now];
+  emitPrices();
+  storage.local
+    .set({ [CACHE_KEY]: priceData })
+    .catch((error) => console.error('Error setting cached cardano token prices', error));
+};
+
+const fetchPrice = async (assetId: Cardano.AssetId) => {
+  const [lastFetchTime, cachedData] = priceData[assetId] || [];
+
+  // If recently fetched, do nothing
+  if (typeof lastFetchTime === 'number' && lastFetchTime > Date.now() - TOKEN_PRICE_CHECK_INTERVAL) return;
+
+  // Immediately set the lastFetchTime to avoid other events fetching the same token price to actually perform the request
+  updatePriceData(assetId, cachedData);
+
+  try {
+    const url = `https://coingecko.live-mainnet.eks.lw.iog.io/api/v3/onchain/networks/cardano/tokens/${assetId}/pools`;
```

Comment:
Consider providing this URL via env variable
```suggestion
    const url = `${process.env.TOKEN_PRICES_URL}/${assetId}/pools`;
```

Reply:
e101d0c61 should solve this